torch/_inductor/codegen/cpp_micro_gemm.py

```diff
@@ -504,18 +504,67 @@ class CppMicroGemmAMX(CppMicroGemm):
     This class generates the code for micro gemm using Advanced Matrix eXtention (AMX)
     instructions available in 4th generation Intel Xeon for compute.
     It supports input types of torch.bfloat16 with fp32 output.
-    TODO(jgong5): support int8 data type.
     """
 
     TEMPLATE_ENTRY = r"""
 {{declare_kernel}} {
     {{kernel.assert_function}}(N % {{block_n}} == 0, "N dimension must be multiple of {{block_n}}");
     {{kernel.assert_function}}(K % 2 == 0, "K dimension must be multiple of 2");
+{%- if input_dtype == torch.bfloat16 and input2_dtype == torch.int8 %}
+    // Create a stack-allocated buffer for tiles of B.
+    // Except maybe for the tail-case, an AMX tile of B has 16x32 BF16 elements
+    // It's possible that N may not be equal to Nr, but might be a multiple of it.
+    // In that case, Kc (which is K in the microkernel) would be smaller (than the case in which N == Nr).
+    const auto num_elements_per_b_tile = 512;
+    const auto buf_size_per_nr_block = ((K + {{block_k}} - 1) / {{block_k}}) * num_elements_per_b_tile * 2;
+    const auto buf_size = buf_size_per_nr_block * (N / {{block_n}});
+    alignas(4096) {{input_t}} bf16_weights_buf[buf_size];
+
+    const auto last_k_offset = K / {{block_k}} * {{block_k}};
+    const auto tail_k_size = K - last_k_offset;
+    const auto num_b_rows = (last_k_offset > 0) ? 16 : (tail_k_size * sizeof({{input_t}})) / 4;
+    const auto b_tile_ptr_stride = ldb * {{vnni_size}};
+
+    auto load_B_row = [&]({{input2_t}}* {{restrict_keyword}} src, {{input_t}}* {{restrict_keyword}} dst) {
+        auto b_int8 = at::vec::Vectorized<int8_t>::loadu(src, static_cast<int64_t>(32));
+        auto b_bf16 = at::vec::convert<{{input_t}}>(b_int8);
+        b_bf16.store(dst);
+    };
+
+    auto load_B_in_buf = [&]({{input2_t}}* B_ptr, int idx) {
+        {{input_t}}* base_addr = bf16_weights_buf + idx;
+        {{kernel.unroll_pragma(8)}}
+        for (int i = 0; i < num_b_rows; i++) {
+            load_B_row(
+                B_ptr + i * b_tile_ptr_stride,
+                base_addr + i * 32
+            );
+        }
+    };
+    auto load_weights = [&](int n) {
+        // Load weights & cache them in L1D.
+        const int64_t init_idx =  n * buf_size_per_nr_block;
+        {{kernel.unroll_pragma(4)}}
+        for (int k = 0; k < last_k_offset; k += {{block_k}}) {
+            {{kernel.unroll_pragma(2)}}
+            for (int tile_col = 0; tile_col <= 1; tile_col++) {
+                load_B_in_buf(const_cast<{{input2_t}}*>(B) + k * ldb + tile_col * {{16 * vnni_size}},
+                              (k / {{block_k // 2}} + tile_col) * num_elements_per_b_tile + init_idx);
+            }
+        }
+    };
+{%- endif %}
     // TODO(jgong5): loop unroll for M and N
     for (int64_t m = 0; m < M; m += {{block_m}}) {
         int64_t block_m = std::min<int64_t>(M - m, {{block_m}});
         int64_t m_tail = m;
         for (int64_t n = 0; n < N; n += {{block_n}}) {
```

Comment:
Perhaps we can move this loop to the outer of "m" loop so that we don't have to check "m == 0" each time.

Reply:
It'll have to be checked for `m == 0`, for each value of `n`. While currently, `n==0`, that may change in the future.
However, UTs can't cover this case, so may I revert to assuming `N == Nr`? Thanks!

Reply:
No, we shouldn't assume `N == Nr`. If we do the following, I suppose we don't need to check `m == 0`?
```
for n ...
   dequantize B
   for m ...
```

Reply:
Thanks, loop-inversion had resulted in perf degradation. But I'll check again

Reply:
Do you know why?

Reply:
No, I don't see it anymore. I'm surprised that I encountered it earlier because even earlier, it shouldn't have made a difference.

Reply:
Inverted the loops, thanks!

Reply:
@jgong5, on second thought, in the future, when `N != Nr`, would it be better for cache-locality if the loop would not be inverted, as A tiles could be reused for same `m`?

Also, in that case, `N` could be made a multiple of `Nr`, and `K` could be decreased, so that just like the code in the main branch, some multiple of `Nr * K` bytes of `B` could be resident in L1D.

Since `N != Nr` case can't be easily tested, I'm not sure about that but will add a comment that the loop inversion may need to be reverted in that case.

Thanks!